### PR TITLE
Remove foreign key constraint on SerializedDagModel's dag_runs field

### LINKED ISSUE #12448: General Error Deleting DAG run
Using a 2.0.0 beta 2 build...

Untitled

Log from webserver:

```
[2020-11-18 14:58:52,712] {interface.py:713} ERROR - Delete record error: Dependency rule tried to blank-out primary key column 'job.id' on instance '<SchedulerJob at 0x7fc68ba7f990>'
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask_appbuilder/models/sqla/interface.py", line 698, in delete
    self.session.commit()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/scoping.py", line 163, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1042, in commit
    self.transaction.commit()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 504, in commit
    self._prepare_impl()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 483, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 2536, in flush
    self._flush(objects)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 2678, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    with_traceback=exc_tb,
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 2638, in _flush
    flush_context.execute()
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 538, in execute
    self.dependency_processor.process_deletes(uow, states)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/dependency.py", line 547, in process_deletes
    state, child, None, True, uowcommit, False
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/dependency.py", line 604, in _synchronize
    sync.clear(dest, self.mapper, self.prop.synchronize_pairs)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/orm/sync.py", line 88, in clear
    "column '%s' on instance '%s'" % (r, orm_util.state_str(dest))
AssertionError: Dependency rule tried to blank-out primary key column 'job.id' on instance '<SchedulerJob at 0x7fc68ba7f990>'
172.19.0.1 - - [18/Nov/2020:14:58:52 +0000] "POST /dagrun/delete/147 HTTP/1.1" 302 415 "http://localhost:8080/dagrun/list/?_flt_3_dag_id=airflow_community_data_refresh&_flt_3_state=success" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.193 Safari/537.36"
```

### PR DESCRIPTION
**Pull request by hsnprsd**
Fixes issue https://github.com/apache/airflow/issues/12448

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by hsnprsd** [comment]
@ashb Correct me if I'm wrong but I think the foreign key constraint is not required as `SerializedDagModel` gets its `dag_id` from `DagModel`. It's just a join relationship for ease of use. It also fixes the issue mentioned above.

**Comment by ashb** [comment]
Every time I add relations I had to look it up. What difference does this make to using this join?

**Comment by ashb** [comment]
Also can you explain in more detail exactly how this fixes the linked issue

**Comment by ashb** [comment]
And we need unit tests to prevent regressions please.

**Comment by hsnprsd** [comment]
@ashb I will add unit tests and explanations soon. Can you tell me which file to put my tests on?

**Comment by ashb** [comment]
Possibly test_dagrun? To ensure we can delete it, even when DagModel and SerialisedDag rows exist?

File: airflow/models/serialized_dag.py [hunk 1]
```diff
@@ -73,8 +73,7 @@ class SerializedDagModel(Base):
 
     dag_runs = relationship(
         DagRun,
-        primaryjoin=dag_id == DagRun.dag_id,
-        foreign_keys=dag_id,
+        primaryjoin=dag_id == foreign(DagRun.dag_id),
```
Comment (kaxil) [inline airflow/models/serialized_dag.py:76]: I am still not sure if we explicitly require `foreign` keyword over here.

Can you explain why it is needed?

based on the 2nd section in https://docs.sqlalchemy.org/en/13/orm/join_conditions.html#creating-custom-foreign-conditions

>An alternative syntax to the above is to use the foreign() and remote() annotations, inline within the relationship.primaryjoin expression. This syntax represents the annotations that relationship() normally applies by itself to the join condition given the relationship.foreign_keys and relationship.remote_side arguments. These functions may be more succinct when an explicit join condition is present, and additionally serve to mark exactly the column that is “foreign” or “remote” independent of whether that column is stated multiple times or within complex SQL expressions:

Comment (hsnprsd) [reply airflow/models/serialized_dag.py:76]: You are right. I tested and we don't need that. I am new to Airflow and SQLAlchemy tbh :).

Comment (kaxil) [reply airflow/models/serialized_dag.py:76]: Looks like the CI failed with following error:

```
sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class SerializedDagModel->serialized_dag'. Original exception was: Could not locate any relevant foreign key columns for primary join condition 'serialized_dag.dag_id = dag_run.dag_id' on relationship SerializedDagModel.dag_runs.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation.
```

So indeed it looks like we do need `foreign`

Comment (kaxil) [reply airflow/models/serialized_dag.py:76]: or maybe it just requires the following:

```
primaryjoin="SerializedDagModel.dag_id == foreign(DagRun.dag_id)",
```

based on example in https://stackoverflow.com/a/57454638/5691525

File: airflow/models/serialized_dag.py [hunk 2]
```diff
@@ -74,7 +74,6 @@ class SerializedDagModel(Base):
     dag_runs = relationship(
         DagRun,
         primaryjoin=dag_id == DagRun.dag_id,
```
Comment (kaxil) [inline airflow/models/serialized_dag.py]:
```suggestion
        primaryjoin=dag_id == foreign(DagRun.dag_id),
```

**Review by kaxil — APPROVED** [review]

**Pull request merged**